_published/0017-azuredevops-repos-and-not-projects.md

```diff
@@ -0,0 +1,259 @@
+that new registry address is writable and wherever that address is referenced
+uses those above built-in variables then will work fine.
+
+The issue is where users have hardcoded these URLs into their continuous
+deployments. The Docker images will be pushed to a new location, but the old
+location is still what's referenced, causing their builds to break.
+
+This information needs to go out, preferably before this is released, to all
+known users of wharf-provider-azuredevops telling them to start using
+`${REPO_GROUP}` and `${REPO_NAME}` instead.
+
+For the unknown users, we should try and word the CHANGELOG.md as clear as
+possible declaring these impacts. We should also write a document on
+<https://iver-wharf.github.io/> about upgrading wharf-provider-azuredevops from
+v1 to v2, where we explain these backward incompatibilities as well.
+
+## Alternative solutions
+
+For the sake of resolving
+<https://github.com/iver-wharf/wharf-provider-azuredevops/issues/24> we could
+settle with still importing 1 Wharf project per Azure DevOps project, but
+this is so tightly coupled so I [@jilleJr] think it's best to tackle this
+inconsistency here and now. This RFC is not meant to fix that bug, but instead
+designed so that bug is fixed as a consequence.
+
+## Future possibilities
+
+- We can update the README.md inside wharf-provider-azuredevops :)
+
+  ```diff
+  -Import Wharf projects from Azure DevOps repositories. Mainly focused on
+  -importing from self hosted Azure DevOps instances, importing from
+  -dev.azure.com is not well tested.
+  +Import Wharf projects from Azure DevOps repositories. Tested on importing
+  +both from self hosted Azure DevOps instances as well as from <dev.azure.com>.
+  ```
+
+- For this issue: [Map projects per ID: update importers code #6](https://github.com/iver-wharf/wharf-provider-azuredevops/issues/6)
+  we can import based on the Azure DevOps repository's GUIDs instead of the
+  project GUIDs, which would be more accurate and would mean less backward
+  compatibility migrations for the future, as this "importing repos and not
+  projects" change seems inevitable anyway.
+
+## Unresolved questions
+
+- How to handle the naming?
+
+  There are three main naming styles:
+
+  | ID  | Project group                 | Project name     |
+  | --- | -------------                 | ------------     |
+  | A   | `{azure-org}/{azure-project}` |`{azure-repo}`    |
+  | B   | `{azure-org}`                 |`{azure-repo}`    |
+  | C   | `{azure-org}`                 |`{azure-project}` |
+
+  The latter, naming C, is used in wharf-provider-azuredevops v1. For v2, how
+  do we handle the naming when there are more than 1 code repository?
+
+  1. Always use the naming A.
+
+  2. Use the naming C, unless the Azure DevOps project contains more than 1
+     code repository, where naming A should be used instead.
+
+  3. Use the naming B/C when the `{azure-repo}` = `{azure-project}`, and use
+     naming A otherwise.
```

Comment:
Could you null check the string and concatenate them such that the project group collumn becomes the name? Id dosent seem like a human friendly name.

Reply:
Sorry for the confusion. The `{var}` fields here represent the different names, and not the IDs/GUIDs. I will add the following snippet that I added to the "Wharf project naming" section:

---

Where:

- `{azure-org}` = name of the Azure DevOps organization that holds the Azure
  DevOps project.

- `{azure-project}` = name of the Azure DevOps project.

- `{azure-repo}` = name of the code repository inside the Azure DevOps project.

Reply:
I don't fully understand what you mean with

> Could you null check the string and concatenate them such that the project group collumn becomes the name?

None of the names of the Azure DevOps org, project, nor repo can be null. The idea was to concatenate the org and the project names (with a slash `/` inbetween) in some cases, to be used as the Wharf group name. Similar to how we have concatenated subgroups in the wharf-provider-gitlab where a project at `https://our-internal-gitlab-instance.example.com/topgroup/subgroup/projectname` would result in the Wharf group name `topgroup/subgroup` and the Wharf project name would resolve to `projectname`.

Reply:
To me it appears the project-group collumn is a always available and unique name whereas the project name collumn does not appear unique. I think the uniqueness is important when viewing from wharf, lack of uniqueness seems like it could cause confusion during project selection.

Why use a slash to concatenate the names if the name is to be part of a URL shouldnt `/` then be designated an illegal character?

Reply:
I'll call you so we can resolve this verbally. I think I might need to rubber-duck this a little as well :)

Reply:
For ref: We cleared up the confusion

- For all our currently supported remote providers, the group can never be null. Explained in detail over at the `REPO_GROUP` env var documentation: https://iver-wharf.github.io/#/usage-wharfyml/variables/built-in-variables?id=repo_group

  > - For GitLab: Namespace (user, group, or group/subgroups)
  > - For AzureDevOps: Organization or collection name.
  > - For GitHub: Organization name or username.

- A future enhancement might be to normalize the Wharf group names so we can have them in the URL without custom/convoluted routing logic. Perhaps by replacing `/` with `-`. Leaving that for a different day though.

- Another future plan, which is referenced in https://github.com/iver-wharf/wharf-provider-azuredevops/issues/6 and https://iver-wharf.github.io/wharf-notes/hide-providers-behind-api is to start using the remote IDs for mapping projects instead of this "by-name basis" that we have now. But that work is already ongoing